File: 1. Two Sum/optimized.js
#!/usr/bin/env node
/**
 * @param {number[]} nums
 * @param {number} target
 * @param {number[]}
 */

var twoSum = function(nums, target) {
	const seen = new Map();
	for (let i = 0; i < nums.length; i++) {
		const dif = target - nums[i];
		if (seen.has(dif)) {
			return [seen.get(dif), i];
		}
		seen.set(nums[i], i);
	}
};

console.log(twoSum([2, 7, 11, 15], 9));
console.log(twoSum([3, 2, 4], 6));
console.log(twoSum([3, 3], 6));

File: 1. Two Sum/optimized.py
#!/usr/bin/env python

from typing import List

def twoSum(nums: List[int], target: int) -> List[int]:
    """Compute indices of two number whose sum is target"""
    seen = {}
    for i, num in enumerate(nums):
        dif = target - num
        if dif in seen:
            return [seen[dif], i]
        seen[num] = i

if __name__ == '__main__':
    print(twoSum([2, 7, 11, 15], 9));
    print(twoSum([3, 2, 4], 6));
    print(twoSum([3, 3], 6));
